### PR TITLE
Do not create iteration variables in foreign bundles

### PR DESCRIPTION
**Pull request by vpodzime**

### CODE REVIEW

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6494/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6494/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6494/

File: libpromises/eval_context.c [hunk 1]
```diff
@@ -1950,24 +1954,65 @@ StringSet *EvalContextStackPromisees(const EvalContext *ctx)
...
+        {
+            new_lval = MangleScopedVarNameIntoSpecialScopeName(ref->scope, lval);
+        }
+        VarRefDestroy(ref);
+    }
     if (strchr(lval, '['))
     {
         // dealing with (legacy) array reference in lval, must parse
-        VarRef *ref = VarRefParseFromScope(lval, SpecialScopeToString(scope));
+        VarRef *ref = VarRefParseFromScope(new_lval ? new_lval : lval, SpecialScopeToString(scope));
```
Comment (larsewi) [inline libpromises/eval_context.c:2003]: CONTRIBUTING.md says to use explicit comparisons. I.e. `new_lval != NULL` instead of ~~`new_lval`~~. Although I'm not sure if this applies here.

File: libpromises/eval_context.c [hunk 2]
```diff
@@ -1950,24 +1954,65 @@ StringSet *EvalContextStackPromisees(const EvalContext *ctx)
...
+        free(new_lval);
         VarRefDestroy(ref);
         return ret;
     }
     else
     {
         // plain lval, skip parsing
-        const VarRef ref = VarRefConst(NULL, SpecialScopeToString(scope), lval);
-        return EvalContextVariablePut(ctx, &ref, value, type, tags);
+        const VarRef ref = VarRefConst(NULL, SpecialScopeToString(scope), new_lval ? new_lval : lval);
```
Comment (larsewi) [inline libpromises/eval_context.c:2012]: Same here

**Review by larsewi — COMMENTED** [review]
Good stuff !

File: libpromises/eval_context.c [hunk 3]
```diff
@@ -1950,24 +1954,65 @@ StringSet *EvalContextStackPromisees(const EvalContext *ctx)
...
+    return new_var_name;
+}
+
 /*
  * Copies value, so you need to free your own copy afterwards.
  */
 bool EvalContextVariablePutSpecial(EvalContext *ctx, SpecialScope scope, const char *lval, const void *value, DataType type, const char *tags)
 {
+    char *new_lval = NULL;
+    if (strchr(lval, '.') != NULL)
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/eval_context.c:1991]: Sure, I didn't know we had this amazingly useful function. :smile:

File: libpromises/eval_context.c (hunk 1 above)
Comment (vpodzime) [reply libpromises/eval_context.c:2003]: I was actually thinking about this. But I think it's just better this way in the tertiary operator because the explicit comparison adds a lot of noise.

Comment (larsewi) [reply libpromises/eval_context.c:2003]: Sure, I agree

**Review by craigcomstock — APPROVED** [review]

File: libpromises/eval_context.c [hunk 4]
```diff
@@ -1950,24 +1954,65 @@ StringSet *EvalContextStackPromisees(const EvalContext *ctx)
...
+    if (StringContainsChar(lval, '.'))
+    {
+        VarRef *ref = VarRefParse(lval);
+        if (ref->scope != NULL)
+        {
+            new_lval = MangleScopedVarNameIntoSpecialScopeName(ref->scope, lval);
+        }
+        VarRefDestroy(ref);
+    }
     if (strchr(lval, '['))
```
Comment (larsewi) [inline libpromises/eval_context.c:2000]: Could use `StringContainsChar` here as well. Although it's old code.

**Review by larsewi — APPROVED** [review]

File: libpromises/eval_context.c (hunk 4 above)
Comment (vpodzime) [reply libpromises/eval_context.c:2000]: Yeah, not worth ruining git history.

**Comment by vpodzime** [comment]
> Alright, I triggered a build:
> 
> [Build Status](https://ci.cfengine.com//job/pr-pipeline/6494/)
> 
> ****

Only an unrelated failure in Selenium tests.

**Review by olehermanse — APPROVED** [review]
Looks good in general, just some comments on potential edge cases.

File: libpromises/eval_context.c [hunk 5]
```diff
@@ -52,6 +52,10 @@
 #include <conversion.h>                               /* DataTypeIsIterable */
 #include <cleanup.h>
 
+/* If we need to put a scoped variable into a special scope, use the string
+ * below to replace the original scope separator.
+ * (e.g. "config.var" -> "this.config___var" ) */
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/eval_context.c:57]: You'd only have problems if you defined variables like that in the bundle and at the same time tried to iterate over `config.var[$list][something]`. I don't think that's practically likely to happen. We could use a different sequence than `___` to make it even more unlikely, but I don't want to forbid users using `___` in the names of their variables.

File: libpromises/eval_context.c [hunk 6]
```diff
@@ -1950,24 +1954,65 @@ StringSet *EvalContextStackPromisees(const EvalContext *ctx)
...
+    char *new_var_name = xmalloc(var_name_len + sizeof(NESTED_SCOPE_SEP));
+    memcpy(new_var_name, var_name, var_name_len + 1 /* including '\0' */);
+
+    /* Make sure we only replace the "scope." string, not all dots. */
+    char *scope_with_dot = StringConcatenate(2, scope, ".");
+    char *scope_with_underscores = StringConcatenate(2, scope, NESTED_SCOPE_SEP);
+
+    NDEBUG_UNUSED ssize_t ret = StringReplace(new_var_name, var_name_len + sizeof(NESTED_SCOPE_SEP),
+                                              scope_with_dot, scope_with_underscores);
+    assert(ret == (var_name_len + sizeof(NESTED_SCOPE_SEP) - 2));
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/eval_context.c:1977]: > `scope.` should only appear at the beginning of the string, right? If so, I'd prefer something like:

I'm not 100% about this so I chose to go the safer way. The `assert()` there is really just to make sure the size calculations are correct.

**Review by olehermanse — APPROVED** [review]
Looks good to me :)
